### DISCUSSION TITLE
Can an injector inject all implementations of an interface?

### DISCUSSION THREAD

**Discussion by zzl221000** [activej/activej#303] [Q&A] [answered]
```java
    interface MyFunc {
        void something();
    }

    @Inject
    class MyFuncImpl1 implements MyFunc {

        @Override
        public void something() {

        }
    }

    @Inject
    class MyFuncImpl2 implements MyFunc {

        @Override
        public void something() {

        }
    }

    @Provides
    public String useInject(Set<MyFunc> myFuncSet) {
        // get all MyFunc implement
        return "";
    }
```

**Comment by eduard-vasinskyi** [comment, marked as answer]
Hi, @zzl221000

An Injector can inject all implementations of a class, but to do so, it needs to [be instructed how to resolve conflicting bindings](https://activej.io/inject/duplicate-bindings).

For your case, the easiest way to do this is by instructing Injector to use Set to resolve conflicting bindings.
You can do this by overriding the `configure()` method of a Module and specifying that you want your bindings to be provided into a set:
```java
bindIntoSet(Key.of(MyFunc.class), Key.of(MyFuncImpl1.class));
bindIntoSet(Key.of(MyFunc.class), Key.of(MyFuncImpl2.class));
```
The resulting class would look something like this:
```java
public class TestModule extends AbstractModule {

   @Override
   protected void configure() {
      bindIntoSet(Key.of(MyFunc.class), Key.of(MyFuncImpl1.class));
      bindIntoSet(Key.of(MyFunc.class), Key.of(MyFuncImpl2.class));
   }


   public interface MyFunc {
      void something();
   }


   @Inject
   static class MyFuncImpl1 implements MyFunc {

      @Override
      public void something() {
         System.out.println("1");
      }
   }


   @Inject
   static class MyFuncImpl2 implements MyFunc {

      @Override
      public void something() {
         System.out.println("2");
      }
   }


   @Provides
   public String useInject(Set<MyFunc> myFuncSet) {
      for (MyFunc myFunc : myFuncSet) {
         myFunc.something();
      }
      // get all MyFunc implement
      return "some string";
   }


   public static void main(String[] args) {
      Injector injector = Injector.of(new TestModule());
      String instance = injector.getInstance(String.class);


      System.out.println(instance);
   }
}
```

When the `main()` method is launched, it prints out:
```bash
2
1
some string
```

**Reply by zzl221000** [reply]
@eduard-vasinskyi 
Thank you very much for the code!

I tried the manual configuration and it solves the problem.

Is there a more convenient way for injector to handle this kind of problem in the future.
For example, the inject annotation does this automatically.

Setting up the MyFunc class to automatically merge into Set
```java
@Inject(alias=MyFunc.class)
class MyFuncImpl1 MyFuncImpl1 implements MyFunc {
      @Override
      public void something() {
         System.out.println("1");
      }
   }
```

**Reply by eduard-vasinskyi** [reply]
@zzl221000

If you do not want to manually configure those bindings, you could replace the `configure()` method with `@ProvidesIntoSet` methods:
```java
@ProvidesIntoSet
MyFunc myFuncImpl1(MyFuncImpl1 myFunc) {
   return myFunc;
}


@ProvidesIntoSet
MyFunc myFuncImpl2(MyFuncImpl2 myFunc) {
   return myFunc;
}
```

An `@Inject` annotation simply creates a new binding for exactly the type it is used on. So you need to let the Injector know that you want to bind interfaces (not actual classes) into a set. You can do it either manually or using reflection (`@ProvidesIntoSet` - methods)
